mlrun/projects/project.py

```diff
@@ -2043,7 +2069,23 @@ def push(self, branch, message=None, update=True, remote=None, add: list = None)
 
         if not branch:
             raise ValueError("please specify the remote branch")
-        repo.git.push(remote or "origin", branch)
+
+        remote = remote or "origin"
+        clean_remote = repo.remotes[remote].url
+        enriched_remote, is_remote_enriched = add_credentials_git_remote_url(
+            clean_remote, secrets=secrets or {}
+        )
+        try:
+            if is_remote_enriched:
+                repo.remotes[remote].set_url(enriched_remote, clean_remote)
+            repo.git.push(remote, branch)
+        except RuntimeError as e:
+            raise mlrun.errors.MLRunRuntimeError(
+                "Failed to push changes to Git repository"
+            ) from e
+        finally:
+            if is_remote_enriched:
+                repo.remotes[remote].set_url(clean_remote, enriched_remote)
```

Comment:
this is also needed for pull. lets extract it to a method and use it for both push or pull. (can use a callback to determine what action is needed)

Reply:
Good thinking. I extracted this logic into a new `_run_authenticated_git_action` method and added its usage to `.pull` as well.